File: 1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:            
        nums.sort()
        l, r = 0, len(nums)
        while l <= r:
            mid = (l + r) >> 1
            cnt = sum(x >= mid for x in nums)
            if cnt == mid:
                return mid
            if cnt < mid:
                r = mid - 1
            else:
                l = mid + 1
        return -1
            

File: 1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var specialArray = function(nums) {
  nums.sort((a, b) => b - a);
  let [l, r] = [0, Math.max(...nums)];
  
  while (l <= r) {
    const mid = (l + r) >>> 1;
    const x = findSpecial(nums, mid);
    if (x === mid) return x;
    
    if (x > mid) l = mid + 1;
    else r = mid - 1;
  }
  
  return -1;
};

const findSpecial = (nums, mid) => nums.reduce((acc, curr) => acc + (curr >= mid), 0);